mesonbuild/interpreter/interpreter.py

```diff
@@ -3381,3 +3383,61 @@ def func_range(self, node, args: T.Tuple[int, T.Optional[int], T.Optional[int]],
         if step < 1:
             raise InterpreterException('step must be >=1')
         return P_OBJ.RangeHolder(start, stop, step, subproject=self.subproject)
+
+    @FeatureNew('response_file', '1.2.0')
+    @typed_pos_args('response_file', varargs=(str, mesonlib.File))
+    @typed_kwargs(
+        'response_file',
+        KwargInfo('prefix', str, default='@'),
+        KwargInfo('quote', (bool, str), default=True),  # TODO: choices...
+        KwargInfo('max_args_length', (bool, int), default=127),  # TODO: sensible default?
+        KwargInfo('separator', str, default=' '),
+        KwargInfo('relative_paths', bool, default=True),
+    )
+    def func_response_file(self, node: mparser.BaseNode, args: T.Tuple[T.Union[str, mesonlib.File, T.List[mesonlib.FileOrString]]], kwargs: kwtypes.ResponseFile) -> T.Union[T.List[str], str]:
+
+        import shlex
+        from ..backend.ninjabackend import cmd_quote, gcc_rsp_quote
+
+        arguments = mesonlib.listify(list(args), flatten=True)
+        if not arguments:
+            return []
+
+        max_args_length = kwargs['max_args_length']
+        if max_args_length is False:
+            return arguments
+        elif max_args_length is True:
+            max_args_length = 0
+
+        quote_fcts = {
+            'sh': shlex.quote,
+            'gcc': gcc_rsp_quote,
+            'cmd': cmd_quote,
+            True: mesonlib.quote_arg,
+        }
+        quote_style = kwargs['quote']
+
+        def quote(value: mesonlib.FileOrString):
+            if isinstance(value, mesonlib.File):
+                if kwargs['relative_paths']:
+                    value = value.relative_name()
+                else:
+                    value = value.absolute_path(self.environment.source_dir, self.environment.build_dir)
+            if quote_style:
+                value = quote_fcts[quote_style](value)
+            return value
+
+        quoted_args = kwargs['separator'].join(quote(a) for a in arguments)
+
+        if len(quoted_args) <= max_args_length:
+            return arguments
+
+        rsp_index = self.rsp_files.setdefault(self.subdir, 0)
+        rsp_file = Path(self.subdir) / f"args{rsp_index}.rsp"
+        self.rsp_files[self.subdir] += 1
+
+        mlog.debug(f"Writing rsp file {rsp_file}")
+        (Path(self.environment.build_dir) / rsp_file).write_text(quoted_args, encoding='utf-8')
```

Comment:
Might not be worth converting to a path?

Reply:
I do not agree on this. `Path` allows to concatenate the path, to open the file, and to write to it. That would require three different functions otherwise. And imho, we should always prefer `Path` over `os.path` manipulations, when this is possible, mostly because it adds the path semantics `str` doen't have.

Reply:
I think it's fine using Path here. Personally I would have used `Path(self.environment.build_dir, rsp_file)` to avoid the extra parentheses, but that's a matter of taste I guess.

Reply:
Or even `rsp_file = Path(self.environment.build_dir, self.subdir, f'args{rsp_index}.rsp')` to make it in one call. BTW, our coding style is to use single quote and you did `f"args{rsp_index}.rsp"` above.